Bhumi guides active record basics (#3)
bhumi1102/rails: guides/source/active_record_basics.md

----------------------

> Active Record is the M in [MVC][] - the model - which is the layer of the system

Author: Yeah I see what you mean. I updated it to actually spell that out and include a reference to Active Model too. 

I think leaving the "M in MVC" part is useful. Especially for a beginner to connect the dots. We do say that Action Controll is the "C in MVC" in that guide. I'm leaning towards adding "V in MVC" to the Action View guide, that I'm currently editing..

> Active Record is the M in [MVC][] - the model - which is the layer of the system

Reviewer: Makes sense to mention "V in MVC" in that guide.

I like the idea of connecting the dots more easily, we can totally keep that, just would like to find ways to disconnect a bit that C is only AR like I mentioned, because has been the common thought for a long time. :+1:

responsible for representing data and business logic. Active Record facilitates
the creation and use of Ruby objects whose attributes require persistent storage
to a database. It is an implementation of the Active Record pattern which itself
is a description of an Object Relational Mapping system.

### The Active Record Pattern

> [Active Record pattern is described by Martin Fowler][MFAR] in the book _Patterns of Enterprise Application Architecture_ as "an object that wraps a row in a database table, encapsulates the database access, and adds domain logic to that data." Active Record objects carry both data and behavior. Active Record pattern encourages putting data access logic in the domain object. This way users of the object know how to write to and read from the database. Active Record classes match very closely to the record structure of the underlying database.

Reviewer 2: +1 to what @Ridhwana says here.

> [Active Record pattern is described by Martin Fowler][MFAR] in the book _Patterns of Enterprise Application Architecture_ as "an object that wraps a row in a database table, encapsulates the database access, and adds domain logic to that data." Active Record objects carry both data and behavior. Active Record pattern encourages putting data access logic in the domain object. This way users of the object know how to write to and read from the database. Active Record classes match very closely to the record structure of the underlying database.

Author: Yeah it is a bit jargon-y. I got that quote directly from that book on my bookself :) That line is also a paraphrase from the book, but I'm removing it! 

I'm also leveling with the reader and saying this is "industry jargon" ORM, RDBS, AR

----------------------

Active Record is the M in [MVC][] - the model - which is the layer of the system
responsible for representing data and business logic. Active Record helps you
create and use Ruby objects whose attributes require persistent storage
to a database.

> Note that it's also possible to model data with Ruby objects that do *not* need to be backed by a database. [Active Model](active_model_basics.html) is used for that. So Active Record and Active Model together are the M in MVC.

Reviewer: Minor suggestion:

```suggestion
Note that it's also possible to model data with Ruby objects that do *not* need to be backed by a database. [Active Model](active_model_basics.html) is commonly used for that in Rails, making Active Record and Active Model both part of the M in MVC, as well as your own plain Ruby objects.
```

And a thought: maybe we move this to an actual note / tip after the first two paragraphs?

> There is some industry jargon around Active Record. Active Record in Rails is an implementation of a software architecture pattern by the same name. It's also a description of something called [Object Relational Mapping][ORM] system. The below sections explain these terms:

Reviewer: Just trying to differentiate the two "Active Record" mentions that are read together:

```suggestion
There is some industry jargon around the term "Active Record". Active Record in Rails is an implementation of a software architecture pattern by the same name. It's also a description of something called an [Object Relational Mapping][ORM] system. The below sections explain these terms:
```

### Active Record as an ORM Framework

Active Record gives us the ability to do the following using Ruby objects:

* Represent models and their data.

> * Represent associations between these models.

Reviewer 3: Totally optional suggestion but I felt it read better without "these" because we not referring to any specific models. 

```suggestion
* Represent associations between models.
```

some naming and schema conventions to follow. And in case you need to, it is possible to [override naming conventions](#overriding-the-naming-conventions).

### Naming Conventions

Active Record uses this naming convention to map between models (represented by
Ruby objects) and database tables:

Rails will pluralize your model's class names to find the respective database
table. For example, a class named `Book` maps to a database table named
**books**. The Rails pluralization mechanisms are very powerful and capable of
pluralizing (and singularizing) both regular and irregular words. This uses the
[Active Support](active_support_core_extensions.html#pluralize) [pluralize
method](https://api.rubyonrails.org/classes/ActiveSupport/Inflector.html#method-i-pluralize),
if you're curious.

For class names composed of two or more words, the model class name will follow
the Ruby conventions of using a CamelCase name. The database table name, in that
case, will be a snake_case name. For example:

* `BookClub` - is the model class, singular with the first letter of each word capitalized
* `book_clubs` - is the matching database table, plural with underscores separating words

> Here are some more examples of model class names and corrosponding table names:

Author: geez, looks like the spell checker wasn't on. sorry about that. fixing all this silly spellings.

| Model / Class    | Table / Schema |

> | ---------------- | -------------- |

Reviewer 3: (optional suggestion, feel free to ignore) I'm not sure if we need so many examples here, maybe one of each type would suffice; Article (for a standard one), LineItem (for snake case), Person (for when the plural does not simply add an s).

Author: haha sure. 'mouse' is out of here.

| `Article`        | `articles`     |
| `LineItem`       | `line_items`   |
| `Product`        | `products`     |
| `Mouse`          | `mice`         |
| `Person`         | `people`       |

### Schema Conventions

Active Record uses conventions for column names in the database tables as well,
depending on the purpose of these columns.

* **Primary keys** - By default, Active Record will use an integer column named
  `id` as the table's primary key (`bigint` for PostgreSQL and MySQL, `integer`
  for SQLite). When using [Active Record

>   Migrations](active_record_migrations.html) to create your tables, this column

Reviewer: Let's maybe link to the migration section within the guide, and let that section point to the separate migrations guide.

  will be automatically created.
* **Foreign keys** - These fields should be named following the pattern
  `singularized_table_name_id` (e.g., `item_id`, `order_id`). These are the
  fields that Active Record will look for when you create associations between
  your models.

There are also some optional column names that will add additional features to
Active Record instances:

* `created_at` - Automatically gets set to the current date and time when the
  record is first created.
* `updated_at` - Automatically gets set to the current date and time whenever
  the record is created or updated.
* `lock_version` - Adds [optimistic
  locking](https://api.rubyonrails.org/classes/ActiveRecord/Locking.html) to a
  model.
* `type` - Specifies that the model uses [Single Table
  Inheritance](https://api.rubyonrails.org/classes/ActiveRecord/Base.html#class-ActiveRecord::Base-label-Single+table+inheritance).
* `(association_name)_type` - Stores the type for [polymorphic
  associations](association_basics.html#polymorphic-associations).
* `(table_name)_count` - Used to cache the number of belonging objects on
  associations. For example, a `comments_count` column in an `Article` class

>   that has many instances of `Comment` will cache the number of existent
>   comments for each article.

Reviewer 3: Should this be existing instead of existent?

```suggestion
  that has many instances of `Comment` will cache the number of existing
  comments for each article.
```

NOTE: While these column names are optional, they are reserved by Active Record.

> Steer clear of reserved keywords when naming your table's columns. For example,

Reviewer 3: I'm wondering if there is a link to somewhere in the guides where have a list of reserved words that would be helpful in this context, if not maybe we can add a task on the board to create one?

Reviewer 2: Good point, and there was a discussion a few years ago about this on the forum: https://discuss.rubyonrails.org/t/reserved-model-names/74559/19 

Added it as [a task](https://3.basecamp.com/3076981/buckets/35498807/todos/7121438912) for now and will let @carlosantoniodasilva weigh in here to see if we should do this (or if it exists somewhere that he is aware of).

Reviewer: I don't think there's a defined list anywhere, it's more about the columns AR uses (which are listed here), but there's things like defining a column named `save` or other AR methods could cause you trouble. 

Rails has some [checks for internal methods](https://github.com/rails/rails/blob/c402ec78724a305d2e22777986aa6f2068d5a0ae/activerecord/lib/active_record/attribute_methods.rb#L30) and it might raise out loud, but it's may not catch everything.

I think it'd be somewhat difficult to keep an up-to-date list of methods, but the columns we have listed here are a pretty good starting point.

FWIW, there's a [similar thing with Action Controller in that guide](https://edgeguides.rubyonrails.org/action_controller_overview.html#methods-and-actions)
![Screenshot 2024-02-26 at 14 37 57](https://github.com/bhumi1102/rails/assets/26328/d2d919b1-6f68-459c-ab4d-e4c234bbb68a)

`type` is a reserved keyword used to designate a table using Single Table
Inheritance (STI). If you are not using STI, use a different word to accurately
describe the data you are modeling.

Creating Active Record Models
-----------------------------

When generating a Rails application, an abstract `ApplicationRecord` class will
be created in `app/models/application_record.rb`. The `ApplicationRecord` class

> inherits from [`ActiveRecord::Base`](https://api.rubyonrails.org/classes/ActiveRecord/Base.html) and it's what turns a regular ruby class into

Reviewer 3: Should Ruby be capitalized?

```suggestion
inherits from [`ActiveRecord::Base`](https://api.rubyonrails.org/classes/ActiveRecord/Base.html) and it's what turns a regular Ruby class into
```

an Active Record model.

`ApplicationRecord` is the base class for all models in your app. To create
Active Record models, subclass the `ApplicationRecord` class and you're good to
go:

```ruby
class Product < ApplicationRecord
end
```

> This will create a `Product` model, mapped to a `products` table in the
> database. Using `ApplicationRecord`, each column in the table is mapped to attributes of the `Product` class. And an instance of the `Product` can represent a row in the `products` table. The `products` table can be created using an SQL statement like this:

Reviewer: We might be able to skip another mention of `ApplicationRecord`:

```suggestion
This will create a `Product` model, mapped to a `products` table in the
database, where each column in the table is mapped to attributes of the `Product` class. An instance of `Product` can represent a row in the `products` table. The `products` table can be created using an SQL statement like this:
```

should be needed only in those cases where you can't follow the convention.

To take advantage of convention over configuration in Active Record, there are
some naming and schema conventions to follow. And in case you need to, it is possible to [override naming conventions](#overriding-the-naming-conventions).

### Naming Conventions

Active Record uses this naming convention to map between models (represented by
Ruby objects) and database tables:

Rails will pluralize your model's class names to find the respective database
table. For example, a class named `Book` maps to a database table named
**books**. The Rails pluralization mechanisms are very powerful and capable of
pluralizing (and singularizing) both regular and irregular words. This uses the
[Active Support](active_support_core_extensions.html#pluralize) [pluralize
method](https://api.rubyonrails.org/classes/ActiveSupport/Inflector.html#method-i-pluralize),
if you're curious.

For class names composed of two or more words, the model class name will follow
the Ruby conventions of using a CamelCase name. The database table name, in that
case, will be a snake_case name. For example:

> * `BookClub` - is the model class, singular with the first letter of each word capitalized
> * `book_clubs` - is the matching database table, plural with underscores separating words

Reviewer: ```suggestion
* `BookClub` - is the model class, singular with the first letter of each word capitalized.
* `book_clubs` - is the matching database table, plural with underscores separating words.
```

* `(association_name)_type` - Stores the type for [polymorphic
  associations](association_basics.html#polymorphic-associations).
* `(table_name)_count` - Used to cache the number of belonging objects on
  associations. For example, a `comments_count` column in an `Article` class
  that has many instances of `Comment` will cache the number of existing
  comments for each article.

NOTE: While these column names are optional, they are reserved by Active Record.
Steer clear of reserved keywords when naming your table's columns. For example,
`type` is a reserved keyword used to designate a table using Single Table
Inheritance (STI). If you are not using STI, use a different word to accurately
describe the data you are modeling.

Creating Active Record Models
-----------------------------

When generating a Rails application, an abstract `ApplicationRecord` class will
be created in `app/models/application_record.rb`. The `ApplicationRecord` class
inherits from [`ActiveRecord::Base`](https://api.rubyonrails.org/classes/ActiveRecord/Base.html) and it's what turns a regular Ruby class into
an Active Record model.

> `ApplicationRecord` is the base class for all models in your app. To create
> Active Record models, subclass the `ApplicationRecord` class and you're good to
> go:

Reviewer: Following that idea of trying to decouple a bit AR = model, maybe we can move `Active Record` to the first sentence:

```suggestion
`ApplicationRecord` is the base class for all Active Record models in your app.
To create a new model, subclass the `ApplicationRecord` class and you're good to
go:
```

);
```

Database tables in Rails are typically created using [Active Record Migrations](#migrations) and not raw SQL. A migration for the `products` table above can be generated like this:

```bash
$ rails generate migration CreateProducts name:string
```

and results in this:

```ruby
# Note: the `id` column, as the primary key, is autometically created by convension. As are `created_at` and `updated_at` columns.

# /db/migrate/20240220143807_create_products.rb

> class CreateProducts < ActiveRecord::Migration

Reviewer 3: This is a totally optional suggestion and not necessary - but I like the idea of using the same example throughout the guide if we can.  For example, we use a "Book" model below and it would be cool to first see how a books table is set up and used, rather than having products table set up and then books be queried.

Again, this is just my preference, I'm curious to hear your thoughts?

Author: This is from the original guide. I created all the Book examples. Will match.

  def change
    create_table :products do |t|
      t.string :name

      t.timestamps
    end
  end
end
```

The SQL as well as the migration above declare a table with two columns: `id`
and `name`. Each row of this table can be represented by an instance of the
`Product` class, with the same two attribures, `id` and `name`. You can access a
product's attributes like this:

```irb

> p = Product.new

Reviewer 3: Should we add the `irb` command prompt like in other places?

```suggestion
irb> p = Product.new
```

Author: Oh right! so I thought about this and I left that out on purpose because it isn't actually `irb`. These examples are from a `rails console` session. In fact, if a reader starts a `irb` REPL session and tries to copy/paste, most of these code snippets won't work. They are rails application specific and not plain ruby.

I was struggling with this so ended up leaving it out. But I know the guidelines say to use `>irb` some place. I can add it back in though wondering if anyone else shares by concern about it being misleading to the reader? cc: @carlosantoniodasilva

Reviewer: I couldn't find mention of this in the guides guidelines, only in the [api guidelines](https://guides.rubyonrails.org/api_documentation_guidelines.html#irb), but it's more about the output than the code.  We do use `irb>` in most Ruby code examples, so maybe for now we stick with it. The copy button does not copy that, so it's helpful there.

That said, I wonder if we should just use `>>` as it'd be what shows up in irb/rails console if not configured with a custom prompt. Perhaps that should be a discussion we could open in Basecamp, to see if it's worth changing across the board.

Reviewer: Apparently the [getting started guide shows the bin/rails console opens up irb](https://edgeguides.rubyonrails.org/getting_started.html#using-a-model-to-interact-with-the-database), and shows the next examples with `irb>`, so let's keep using that.

=> #<Product:0x00007f95c7ca94f0 id: nil, name: nil, created_at: nil, updated_at: nil>

> p.name = "pen"
> => "pen"
> p.name
> => "pen"

Reviewer 3: ```suggestion
irb> p.name = "pen"
=> "pen"
irb> p.name
=> "pen"
```

the `create` method call will create and save a new record into the database:

```ruby
book = Book.create(title: "The Lord of the Rings", author: "J.R.R. Tolkien")
```

While the `new` method will instantiate an object without saving it to the database:

```ruby
book = Book.new
book.title = "The Hobbit"
book.author = "J.R.R. Tolkien"

# The above `book` is not yet saved to the database

book.save

# Now the `book` record is committed to the database.
```

Finally, if a block is provided, both `create` and `new` will yield the new
object to that block for initialization, while only `create` will persist
the resulting object to the database:

```ruby
book = Book.new do |b|
  b.title = "Metaprogramming Ruby 2"
  b.author = "Paolo Perrotta"
end

book.save
```

The resulting SQL statement from both `book.save` and `Book.create` look something like this:

```sql
INSERT INTO "books" ("title", "author", "created_at", "updated_at") VALUES (?, ?, ?, ?) RETURNING "id"  [["title", "Metaprogramming Ruby 2"], ["author", "Paolo Perrotta"], ["created_at", "2024-02-22 20:01:18.469952"], ["updated_at", "2024-02-22 20:01:18.469952"]]

> ```

Reviewer: It just occurred to me that we could maybe add a note here to mention how `created_at` and `updated_at` got set automatically, since that was kinda mentioned in the migration block... do you think that would add to the example?

Same for `updated_at` in the update subsection/example.

Not sure we'd need to mention about the id. Also feel free to skip/ignore it, if it doesn't seem like a good addition.

Author: Yeah I think that would. I added a little comment in the sql block for both `INSERT` and `UPDATE`.

```ruby
# Return a collection with all books
books = Book.all

# Return a single book
first_book = Book.first
last_book = Book.last
random_book = Book.take
```

The above results in the following SQL:

```sql
# Book.all
SELECT "books".* FROM "books"

# Book.first
SELECT "books".* FROM "books" ORDER BY "books"."id" ASC LIMIT ?  [["LIMIT", 1]]

# Book.last
SELECT "books".* FROM "books" ORDER BY "books"."id" DESC LIMIT ?  [["LIMIT", 1]]

# Book.take
SELECT "books".* FROM "books" LIMIT ?  [["LIMIT", 1]]
```

We can also find specific books with `find_by` and `where`:

```ruby
# Return the first book by the given author
book = Book.find_by(author: "J.R.R. Tolkien")

# Shorthand for Book.find_by(id: 42)
book = Book.find(42)
```

The above resulting in this SQL:

```sql
SELECT "books".* FROM "books" WHERE "books"."author" = ? LIMIT ?  [["author", "J.R.R. Tolkien"], ["LIMIT", 1]]

SELECT "books".* FROM "books" WHERE "books"."id" = ? LIMIT ?  [["id", 42], ["LIMIT", 1]]
```

```ruby
# Find all books with a given an author, sort by created_at in reverse chronological order

> hitchhiker_series = Book.where(author: "Douglas Adams").order(created_at: :desc)

Reviewer: Nit: maybe omitting the variable would shorten the example a bit? I'm not sure how much it adds to the example, and it is not used elsewhere in the next examples.

```suggestion
Book.where(author: "Douglas Adams").order(created_at: :desc)
```

book.update(title: "The Lord of the Rings: The Fellowship of the Ring")
```

the `update` results in the following SQL:

```sql
 UPDATE "books" SET "title" = ?, "updated_at" = ? WHERE "books"."id" = ?  [["title", "The Lord of the Rings: The Fellowship of the Ring"], ["updated_at", "2024-02-22 20:51:13.487064"], ["id", 104]]
```

This is useful when updating several attributes at once. Similar to `create`, using `update` will commit the updated records to the database.

If you'd like to update several records in bulk **without callbacks or
validations**, you can update the database directly using `update_all`:

```ruby
Book.update_all status: "already own"
```

### Delete

Likewise, once retrieved, an Active Record object can be destroyed, which removes it from the database.

```ruby
book = Book.find_by(author: "J.R.R. Tolkien")
book.destroy
```

The `destroy` results in this SQL:

```sql
DELETE FROM "books" WHERE "books"."id" = ?  [["id", 104]]
```

If you'd like to delete several records in bulk, you may use `destroy_by`
or `destroy_all` method:

```ruby
# Find and delete all books by Douglas Adams
Book.destroy_by(author: "Douglas Adams")

# Delete all books
Book.destroy_all
```

> Validations

Reviewer: I think the current examples under these following sections are fine, but would it make sense to use `Book` instead of `User` with them as well?

Author: That's odd, my diff already has `Book` for destroy section. Something might be outdated.

For the top level sections like Validations, I think a little variation can be okay. But within a section like `CRUD` keeping it all to `Book` makes a lot of sense.

# Now the `book` record is committed to the database.
```

Finally, if a block is provided, both `create` and `new` will yield the new
object to that block for initialization, while only `create` will persist
the resulting object to the database:

```ruby
book = Book.new do |b|
  b.title = "Metaprogramming Ruby 2"
  b.author = "Paolo Perrotta"
end

book.save
```

The resulting SQL statement from both `book.save` and `Book.create` look something like this:

```sql
INSERT INTO "books" ("title", "author", "created_at", "updated_at") VALUES (?, ?, ?, ?) RETURNING "id"  [["title", "Metaprogramming Ruby 2"], ["author", "Paolo Perrotta"], ["created_at", "2024-02-22 20:01:18.469952"], ["updated_at", "2024-02-22 20:01:18.469952"]]
```

### Read

Active Record provides a rich API for accessing data within a database. You can query a single record or multiple records, filter them by any attribute, order them, group them, select specific fields, and do anything you can do with SQL.

```ruby
# Return a collection with all books
books = Book.all
```

```ruby
# Return a single book
first_book = Book.first
last_book = Book.last
random_book = Book.take
```

The above results in the following SQL:

> ```sql

Reviewer: Nit: there's no SQL example for `Book.all`, maybe could combine those two Ruby blocks, and add one line here for the resulting SQL query as well, for completeness?

Author: Good point. Added SQL for `Book.all`. I wanted to keep the SQL block separate, as a way to show that the user doesn't need to worry about it to use Active Record. As in "FYI, this is the SQL that's generated behind the scenes."

# Book.first
SELECT "books".* FROM "books" ORDER BY "books"."id" ASC LIMIT ?  [["LIMIT", 1]]

# Book.last
SELECT "books".* FROM "books" ORDER BY "books"."id" DESC LIMIT ?  [["LIMIT", 1]]

# Book.take
SELECT "books".* FROM "books" LIMIT ?  [["LIMIT", 1]]
```

We can also find specific books with `find_by` and `where`:

```ruby
# Return the first book by the given author
book = Book.find_by(author: "J.R.R. Tolkien")

# Shorthand for Book.find_by(id: 42)
book = Book.find(42)
```

The above resulting in this SQL:

```sql
SELECT "books".* FROM "books" WHERE "books"."author" = ? LIMIT ?  [["author", "J.R.R. Tolkien"], ["LIMIT", 1]]

SELECT "books".* FROM "books" WHERE "books"."id" = ? LIMIT ?  [["id", 42], ["LIMIT", 1]]
```

```ruby
# Find all books with a given an author, sort by created_at in reverse chronological order
hitchhiker_series = Book.where(author: "Douglas Adams").order(created_at: :desc)
```

resulting in this SQL:

```sql
SELECT "books".* FROM "books" WHERE "books"."author" = ? ORDER BY "books"."created_at" DESC [["author", "Douglas Adams"]]
```

> There are many more Active Record methods to read and query record. You can learn all about them in the [Active Record Query Interface](active_record_querying.html) guide.

Reviewer 3: ```suggestion
There are many more Active Record methods to read and query records. You can learn more about them in the [Active Record Query Interface](active_record_querying.html) guide.
```

```sql
 UPDATE "books" SET "title" = ?, "updated_at" = ? WHERE "books"."id" = ?  [["title", "The Lord of the Rings: The Fellowship of the Ring"], ["updated_at", "2024-02-22 20:51:13.487064"], ["id", 104]]
```

This is useful when updating several attributes at once. Similar to `create`, using `update` will commit the updated records to the database.

If you'd like to update several records in bulk **without callbacks or
validations**, you can update the database directly using `update_all`:

```ruby
Book.update_all status: "already own"
```

### Delete

Likewise, once retrieved, an Active Record object can be destroyed, which removes it from the database.

```ruby
book = Book.find_by(author: "J.R.R. Tolkien")
book.destroy
```

The `destroy` results in this SQL:

```sql
DELETE FROM "books" WHERE "books"."id" = ?  [["id", 104]]
```

If you'd like to delete several records in bulk, you may use `destroy_by`
or `destroy_all` method:

```ruby
# Find and delete all books by Douglas Adams
Book.destroy_by(author: "Douglas Adams")

# Felete all books
Book.destroy_all
```

Validations
-----------

Active Record allows you to validate the state of a model before it gets written

> into the database. There are several methods that allow for diffrent types of validations. For example, validate that an attribute value is not empty, is unique, is not already in the database, follows a specific format, and many more.

Reviewer 3: ```suggestion
into the database. There are several methods that allow for different types of validations. For example, validate that an attribute value is not empty, is unique, is not already in the database, follows a specific format, and many more.
```

---------

Active Record callbacks allow you to attach code to certain events in the

> life-cycle of your models. This enables you to add behavior to your models by executing code when those events occur, like when you create a new record, update it, destroy it, and so on.

Reviewer: Nit: In Active Model callbacks we used `lifecycle`, and other places are `life cycle`. Maybe we stick to one (at least in relation to callbacks) for consistency? (cc @Ridhwana)

Author: I'll go with "lifecycle"
